pkg/asset/machines/openstack/machines.go

```diff
@@ -37,24 +37,46 @@ func Machines(clusterID string, config *types.InstallConfig, pool *types.Machine
 	if poolPlatform := pool.Platform.Name(); poolPlatform != openstack.Name {
 		return nil, fmt.Errorf("non-OpenStack machine-pool: %q", poolPlatform)
 	}
-	platform := config.Platform.OpenStack
 
-	az := ""
-	provider, err := generateProvider(clusterID, platform, pool.Platform.OpenStack, osImage, az, role, userDataSecret)
+	mpool := pool.Platform.OpenStack
+	platform := config.Platform.OpenStack
+	trunkSupport, err := checkNetworkExtensionAvailability(platform.Cloud, "trunk")
 	if err != nil {
 		return nil, err
 	}
 
-	if role == "master" {
-		provider.ServerGroupName = clusterID + "-master"
-	}
-
 	total := int64(1)
 	if pool.Replicas != nil {
 		total = *pool.Replicas
 	}
 	machines := make([]machineapi.Machine, 0, total)
+	providerConfigs := map[string]*openstackprovider.OpenstackProviderSpec{}
 	for idx := int64(0); idx < total; idx++ {
+		zone := mpool.Zones[int(idx)%len(mpool.Zones)]
+		var provider *openstackprovider.OpenstackProviderSpec
+
+		if _, ok := providerConfigs[zone]; !ok {
+			provider, err = generateProvider(
+				clusterID,
+				platform,
+				mpool,
+				osImage,
+				zone,
+				role,
+				userDataSecret,
+				trunkSupport,
+			)
+			if err != nil {
+				return nil, err
+			}
+			providerConfigs[zone] = provider
+		}
```

Comment:
I do not understand why we are using this map? is this some optimization??

Reply:
yeah its so we only create one install config per Zone